File: fsharp/space-age/SpaceAge.fs
module SpaceAge

type Planet = 
  | Earth
  | Mercury
  | Venus
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

let earthYearInSeconds = 31557600m

let getOrbitalPeriodInEarthYears (planet: Planet): decimal =
  match planet with
    | Earth   -> 1.00000000m
    | Mercury -> 0.24084670m
    | Venus   -> 0.61519726m
    | Mars    -> 1.88081580m
    | Jupiter -> 11.8626150m
    | Saturn  -> 29.4474980m
    | Uranus  -> 84.0168460m
    | Neptune -> 164.791320m

let spaceAge (planet: Planet) (seconds: decimal): decimal =
  let earthYears = seconds / earthYearInSeconds
  let planetYears = earthYears / getOrbitalPeriodInEarthYears planet
  System.Math.Round(planetYears, 2)

File: javascript/space-age/space-age.js
export function age(planet, seconds) {
  return round(seconds / orbitalPeriodInSeconds(planet));
}

function orbitalPeriodInSeconds(planet) {
  if (planet === 'earth')
    return 31557600;

  if (planet === 'mercury')
    return orbitalPeriodInSeconds('earth') * 0.2408467;

  if (planet === 'venus')
    return orbitalPeriodInSeconds('earth') * 0.61519726;

  if (planet === 'mars')
    return orbitalPeriodInSeconds('earth') * 1.8808158;

  if (planet === 'jupiter')
    return orbitalPeriodInSeconds('earth') * 11.862615;

  if (planet === 'saturn')
    return orbitalPeriodInSeconds('earth') * 29.447498;

  if (planet === 'uranus')
    return orbitalPeriodInSeconds('earth') * 84.016846;

  if (planet === 'neptune')
    return orbitalPeriodInSeconds('earth') * 164.79132;

  throw new Error("Unexpected planet");
}

function round(number, decimalDigits) {
  if (decimalDigits == null)
    decimalDigits = 2;

  const power = Math.pow(10, decimalDigits);
  return Math.round(number * power) / power;
}